docs/develop/compiler.rst

```diff
@@ -0,0 +1,417 @@
+        if (s->emit_options == MP_EMIT_OPT_ASM) {
+            // inline assembly
+            if (comp->emit_inline_asm == NULL) {
+                comp->emit_inline_asm = ASM_EMITTER(new)(max_num_labels);
+            }
+            comp->emit = NULL;
+            comp->emit_inline_asm_method_table = ASM_EMITTER_TABLE;
+            compile_scope_inline_asm(comp, s, MP_PASS_CODE_SIZE);
+            #if MICROPY_EMIT_INLINE_XTENSA
+            // Xtensa requires an extra pass to compute size of l32r const table
+            // TODO this can be improved by calculating it during SCOPE pass
+            // but that requires some other structural changes to the asm emitters
+            #if MICROPY_DYNAMIC_COMPILER
+            if (mp_dynamic_compiler.native_arch == MP_NATIVE_ARCH_XTENSA)
+            #endif
+            {
+                compile_scope_inline_asm(comp, s, MP_PASS_CODE_SIZE);
+            }
+            #endif
+            if (comp->compile_error == MP_OBJ_NULL) {
+                compile_scope_inline_asm(comp, s, MP_PASS_EMIT);
+            }
+        } else
+
+The inline assembler code comprises assembly instructions in a Python function.
+See `the inline assembler tutorial 
+<https://docs.micropython.org/en/latest/pyboard/tutorial/assembler.html#pyboard-tutorial-assembler>`_ 
+for more details.
+
+Fourth, fifth and sixth passes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The other two passes compute the stack and code size, while the last pass emits the final code:
+
+.. code-block:: c
+   
+   compile_scope(comp, s, MP_PASS_STACK_SIZE);
+
+   if (comp->compile_error == MP_OBJ_NULL) {
+        compile_scope(comp, s, MP_PASS_CODE_SIZE);
+   }
+
+   if (comp->compile_error == MP_OBJ_NULL) {
+    compile_scope(comp, s, MP_PASS_EMIT);
+   }
+
+Before these passes, there is a selection for the type of code to be emitted which can either be native or
+bytecode.
+
+.. code-block:: c
+
+   switch (s->emit_options) {
+
+        #if MICROPY_EMIT_NATIVE
+        case MP_EMIT_OPT_NATIVE_PYTHON:
+        case MP_EMIT_OPT_VIPER:
+            if (emit_native == NULL) {
+                emit_native = NATIVE_EMITTER(new)(&comp->compile_error, &comp->next_label, max_num_labels);
+            }
+            comp->emit_method_table = NATIVE_EMITTER_TABLE;
+            comp->emit = emit_native;
+            break;
+        #endif // MICROPY_EMIT_NATIVE
+
+        default:
+            comp->emit = emit_bc;
+            #if MICROPY_EMIT_NATIVE
+                comp->emit_method_table = &emit_bc_method_table;
+            #endif
+            break;
+   }
+
+The bytecode option is the default but something unique to note for the native code option is that there is another
+option via ``VIPER``. See the :ref:`Emitting native code <emitting_native_code>` section for
+more details on viper annotations.
+
+Emitting bytecode
+-----------------
+
+For every statement in the code, there is a corresponding function to emit MicroPython bytecode. 
```

Comment:
that's technically not exactly true, and is a little bit misleading... every grammar node defined by `DEF_RULE` has a corresponding compile function in `py/compile.c`, but that doesn't necessarily always lead to emitted bytecode (eg the `nonlocal x` statement emits nothing).

maybe something like this:
```
Statements in Python code usually correspond to emitted bytecode, for example `a + b`
generates ing but instead affect other things like the scope of variables, for example
`global a`.
```